### DISCUSSION TITLE
Conflicting local and external package names leading to ModuleNotFoundError

### DISCUSSION THREAD

**Discussion by HakanSunay** [marcelotduarte/cx_Freeze#1882] [Q&A] [answered]
**TL;DR**
Indirect name duplication in local and external module lead to ModuleNotFoundError.
I want to bypass this and have both the local and external module neatly packaged in `library.zip`.

What is the recommended way to work around an issue involving duplicate names when building binaries with cx_freeze?
I have downloaded `cryptography-40.0.2-pp39-pypy39_pp73-macosx_10_12_x86_64.whl` from https://pypi.org/project/cryptography/#files and unzipped it into a local folder named `deps`.

Now, I want to package all of its content into `library.zip`. However, at the same time, I have a local package named `utils`, which has a couple of python modules within it (I am using `utils/platforms.py` in my `main.py`). The problem is that `cryptography` also has a module named `utils.py`:
```
❯ ls deps/cryptography
__about__.py  __init__.py   exceptions.py fernet.py     hazmat        py.typed      utils.py      x509

❯ ls utils
__init__.py  platforms.py
```
The presence of the above `cryptography/utils.py` file causes cx_freeze to not package my local package:
```
❯ ./build/exe.macosx-10.9-universal2-3.9/main
sys.path is ['/Users/hhalil/test-py/src/build/exe.macosx-10.9-universal2-3.9/lib/library.zip', '/Users/hhalil/test-py/src/build/exe.macosx-10.9-universal2-3.9/lib']
Traceback (most recent call last):
  File "/Users/hhalil/test-py/lib/python3.9/site-packages/cx_Freeze/initscripts/__startup__.py", line 135, in run
    module_init.run(name + "__main__")
  File "/Users/hhalil/test-py/lib/python3.9/site-packages/cx_Freeze/initscripts/console.py", line 18, in run
    exec(code, module_main.__dict__)  # pylint: disable=exec-used
  File "main.py", line 14, in <module>
    import utils.platforms
ModuleNotFoundError: No module named 'utils.platforms'; 'utils' is not a package
```

This is how my `setup.py` looks:
```
import sys
from cx_Freeze import setup, Executable

# causes the utils problem:
sys.path.insert(0, "/Users/hhalil/test-py/src/deps/cryptography")

# causes the cryptography problem:
# sys.path.append("/Users/hhalil/test-py/src/deps/cryptography")

setup(
    name="print_syspath",
    version="1.0",
    executables=[Executable("main.py")],
    options={
        "build_exe": {
            "path": sys.path
        }
    },
)
```

I have informally found out that the first occurrence of "utils" is the one that makes it into the zip.
Is this a known issue? If yes, what is the right way to handle such issues?

Ideally, I want the zip (library.zip) to contain both `cryptography/**` and my local package `utils/**`.

Essentially, I might be looking to resolve the naming conflict when using a pre-downloaded `.whl` which has already been unzipped into a certain directory.

**Comment by HakanSunay** [comment, marked as answer]
The issue was fixed by specifying the right path to cryptography.

I was previously using the internal directory which appears after unzipping the wheel. In other words, if you have:
```
❯ ll deps/cryptography
total 5112
drwxr-xr-x   5 hhalil  staff      160 May  3 17:02 .
drwxr-xr-x   3 hhalil  staff       96 May  3 17:02 ..
drwxr-xr-x@ 10 hhalil  staff      320 May  3 17:02 cryptography
-rw-r--r--@  1 hhalil  staff  2616230 May  3 16:59 cryptography-40.0.2-pp39-pypy39_pp73-macosx_10_12_x86_64.whl
drwxr-xr-x@ 10 hhalil  staff      320 May  3 17:02 cryptography-40.0.2.dist-info
```

Do **NOT use** the nested one like:
```
sys.path.append("deps/cryptography/cryptography")
```
instead **DO use** the parent one like:
```
sys.path.append("deps/cryptography")
```

Keep in mind that the so-called parent one was created by me and inside of it, I unzipped the wheel.
I assume cx_freeze needs access to both the `.dist-info` and the actual bits of the external package.
